src/main/java/com/github/javabdd/JFactory.java

```diff
@@ -2377,6 +2425,974 @@ int zdiff_rec(int l, int r) {
         return res;
     }
 
+    int bdd_relnext(int states, int relation, int vars) {
+        // Check validity of BDD nodes.
+        CHECKa(states);
+        CHECKa(relation);
+        CHECKa(vars);
+
+        // Initialize caches if needed.
+        if (applycache == null) {
+            applycache = BddCacheI_init(cachesize);
+        }
+        if (itecache == null) {
+            itecache = BddCacheI_init(cachesize);
+        }
+
+        // We may also apply the OR operation while computing 'relnext'.
+        applyop = bddop_or;
+
+        int result;
+        int numReorder = 1;
+
+        again:
+        for (;;) {
+            try {
+                INITREF();
+
+                if (numReorder == 0) {
+                    bdd_disable_reorder();
+                }
+                result = relnext_rec(states, relation, vars);
+
+                if (numReorder == 0) {
+                    bdd_enable_reorder();
+                }
+            } catch (ReorderException x) {
+                bdd_checkreorder();
+                numReorder--;
+                continue again;
+            }
+            break;
+        }
+
+        checkresize();
+        return result;
+    }
+
+    int relnext_rec(int states, int relation, int vars) {
+        if (VERIFY_ASSERTIONS) {
+            _assert(!ZDD);
+        }
+
+        if (cachestats.enabled) {
+            cachestats.opAccess++;
+        }
+
+        // Terminals cases.
+        if (ISZERO(states) || ISZERO(relation)) {
+            return bddfalse;
+        }
+        if (ISONE(states) && ISONE(relation)) {
+            return bddtrue;
+        }
+        if (ISCONST(vars)) {
+            return states;
+        }
+
+        // Determine the top level of 'states' and 'relation'.
+        int level_states = LEVEL(states);
+        int level_relation = LEVEL(relation);
+        int level = level_states < level_relation ? level_states : level_relation;
+
+        // Skip levels in the variable set until we reach (or exceed) 'level'.
+        boolean sameHeight = false;
+
+        for (;;) {
+            int level_vars = LEVEL(vars);
+
+            if (level == level_vars || (level ^ 1) == level_vars) {
+                sameHeight = true;
+                break;
+            }
+
+            if (level < level_vars) {
+                break;
+            }
+
+            vars = HIGH(vars);
+
+            if (ISCONST(vars)) {
+                return states;
+            }
+        }
+
+        // Consult the operation cache.
```

Comment:
missing section comment for done with consulting cache, but now doing the next thing (the actual operation?).